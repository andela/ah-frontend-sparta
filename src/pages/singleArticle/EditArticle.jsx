import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import TagList from '../../helpers/TagList';
import EditArticleComponent from '../../components/Articles/editArticle';
import getArticleTags from '../../helpers/handleArticleTags';
import { getSingleArticle, editSingleArticle } from '../../actions/articlesActions';


const KeyCodes = {
  comma: 188,
  enter: 13,
};

const delimiters = [KeyCodes.comma, KeyCodes.enter];

export class EditArticle extends React.Component {
  constructor() {
    super();
    this.state = {
      tags: [
      ],
      title: '',
      description: '',
      body: '',
      Articleslug: '',
    };

    this.handleDelete = this.handleDelete.bind(this);
    this.handleAddition = this.handleAddition.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    const { getArticle } = this.props;
    const { slug } = this.props.match.params;
    this.setState({
      Articleslug: slug,
    });
    getArticle(slug);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps !== this.props) {
      const {
        title, description, body, tags,
      } = nextProps.article.article;
      const tagsArray = [];
      tags.map(tag => tagsArray.push({ id: tag, text: tag }));
      this.setState({
        tags: tagsArray,
      });
      this.setState({
        title, description, body,
      });
    }
  }

  onSubmit = (event) => {
    event.preventDefault();
    const { updateArticle, history } = this.props;
    const articleTags = this.state.tags;
    const tags = getArticleTags(articleTags);
    const { title, description, body } = this.state;
    updateArticle({
      title, description, body, tags,
    }, this.props, this.state.Articleslug);
  }

    handleChange = (event) => {
      event.preventDefault();
      this.setState({
        [event.target.name]: event.target.value,
      });
    }

    handleDelete(i) {
      const { tags } = this.state;
      this.setState({
        tags: tags.filter((tag, index) => index !== i),
      });
    }

    handleAddition(tag) {
      this.setState(state => ({ tags: [...state.tags, tag] }));
    }


    render() {
      const { title, description, body } = this.state;
      const { tags } = this.state;
      const suggestedTags = TagList.map(tag => ({
        id: tag,
        text: tag,
      }));
      return (
        <EditArticleComponent
          tags={tags}
          delimiters={delimiters}
          suggestions={suggestedTags}
          handleDelete={this.handleDelete}
          handleAddition={this.handleAddition}
          onSubmit={this.onSubmit}
          handleChange={this.handleChange}
          title={title}
          description={description}
          body={body}
        />
      );
    }
}

EditArticle.defaultProps = {
  slug: '',
  article: PropTypes.shape(PropTypes.object.isRequired),
  match: PropTypes.shape(PropTypes.object.isRequired),
};

EditArticle.propTypes = {
  getArticle: PropTypes.func.isRequired,
  updateArticle: PropTypes.func.isRequired,
  slug: PropTypes.string,
  article: PropTypes.shape(PropTypes.object.isRequired),
  match: PropTypes.shape(PropTypes.object.isRequired),
};

const mapStateToProps = state => ({
  article: state.article,
});
export default connect(mapStateToProps,
  { getArticle: getSingleArticle, updateArticle: editSingleArticle })(EditArticle);
