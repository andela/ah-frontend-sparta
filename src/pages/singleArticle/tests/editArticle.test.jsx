import React from 'react';
import { shallow, mount } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import { EditArticle } from '../EditArticle';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const initialState = {
  articlesReducer:
        { error: undefined, article: undefined, articles: [] },
};
const store = mockStore(initialState);


describe('Edit Article Page Component', () => {
  let mountApp;
  it('Should render without crashing', () => {
    const props = {
      handleChange: jest.fn(),
      getArticle: jest.fn(),
      updateArticle: jest.fn(),
      match: {
        params: {
          slug: 'new-slug',
        },
      },
    };
    mountApp = mount(
      <Router>
        <Provider store={store}>
          <EditArticle {...props} />
        </Provider>
      </Router>,
    );
    expect(mountApp).toMatchSnapshot();
  });

  it('componentWillReceiveProps', () => {
    const props = {
      handleChange: jest.fn(),
      getArticle: jest.fn(),
      updateArticle: jest.fn(),
      match: {
        params: {
          slug: 'new-slug',
        },
      },
    };
    const wrapper = shallow(<EditArticle {...props} />);
    wrapper.instance().setState = jest.fn();
    wrapper.setProps({
      article: {
        article: {
          id: 23,
          author: {
            username: 'Daniel Roy',
            firstname: null,
            lastname: null,
            bio: null,
            image: null,
            followers_no: 0,
            following_no: 0,
          },
          article_read_time: '4 minute read',
          favorite: [],
          title: 'The Valley in the darkness',
          description: 'A short story by Danny',
          slug: 'the-valley-in-the-darkness-22',
          body: 'come on',
          tags: [
            'technology',
            'Andela',
          ],
        },
      },
    });
    expect(wrapper.instance().setState).toBeCalled();
  });

  it('should call an onChange function', () => {
    const props = {
      updateArticle: jest.fn(),
      getArticle: jest.fn(),
      match: {
        params: {
          slug: 'new-slug',
        },
      },
    };
    const event = {
      preventDefault: () => { },
      target: {
        name: 'body',
        value: 'abc',
      },
    };
    const wrapper = mount(
      <EditArticle {...props} />,
    );
    const bodyInput = wrapper.find('#description').first();
    bodyInput.simulate('change', event);
    expect(event.target.value).toEqual('abc');
  });

  it('should call a submit a form', () => {
    const props = {
      updateArticle: jest.fn(),
      getArticle: jest.fn(),
      match: {
        params: {
          slug: 'new-slug',
        },
      },
    };
    const wrapper = mount(
      <EditArticle {...props} />,
    );
    const instance = wrapper.instance();
    instance.onSubmit({ preventDefault: jest.fn() });
    expect(props.updateArticle).toHaveBeenCalled();
  });
});
