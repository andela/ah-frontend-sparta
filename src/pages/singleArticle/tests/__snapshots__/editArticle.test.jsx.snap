// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Article Page Component Should render without crashing 1`] = `ReactWrapper {}`;
